rfc: configuration and shared files

JP-Ellis opened the pull request:
RFC proposal to try and standardise the way we configure and share files across the Pact ecosystem.

Rendered view can be found [here](https://github.com/pact-foundation/roadmap/blob/rfc/configuration-and-shared-files/rfc/0000-configuration-and-shared-files.md).

Excerpt from the RFC:

> ## Summary
> 
> This RFC proposes to establish a standard way to configure and share common files across the Pact ecosystem, including the Pact Broker, PactFlow, and the various Pact implementations.
> 
> ## Motivation
> 
> The Pact ecosystem is quite diverse and we would like to ensure that transitioning between different tools is as seamless as possible. Several de-facto standards have emerged, but these have not been formalised. Furthermore, there is no clear way to extend these de-facto standards to new tools or use cases.
> 
> This RFC aims to formalise some existing de-facto standards. Furthermore, this RFC proposes new standards with the hope of ensuring a more consistent experience across the Pact ecosystem and into the future.
> 
> This RFC is split into three main categories:
> 
> - **Environment Variables**: Define a standard way to configure Pact implementations using environment variables.
> - **Configuration**: Define a standard way to store runtime configuration for Pact implementations.
> - **Shared Files**: Define a standard way to store shared files across the Pact ecosystem.
>

YOU54F commented:
Looks good to me. I assume for most implementors of client languages, this will be handled by the FFI, and other consumers would be the rust and ruby suite of CLI tools?

YOU54F commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,264 @@
+
+  If the operating system cannot be detected, the fallback location _should_ be used.
+
+  If the operating system is detected and the configuration file is found in the fallback location, a warning _should_ be displayed informing the user to relocate the configuration.
```
should a warning appear, if the configuration file is found in the fallback location, if the user explicitly selected that location, for `PACT_CONFIG_FILE`.

Same q for 

>- `PACT_DATA_HOME`
>- Fallback: `~/.pact/`

> If the operating system is detected and the fallback location is found, a warning _should_ be displayed to inform the user to relocate the directory.


If I am on a supported system, and I set the values, to the same as the fallbacks, `~/.pact/config.toml` & `~/.pact`, should the warning be suppressed?

JP-Ellis replied:
This is indeed ambiguous. The short answer is no, I don't think a warning should not be raised if the configuration (either set through environment variables, or from parsing the config file) points to a fallback location.

YOU54F replied:
thanks dude

YOU54F commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,264 @@
+      2. User configuration file
+3. Default values
+
+The explicit configuration file may either be set by the `PACT_CONFIG_FILE` environment variable, or by a command line option.
```
Where does a CLI provided option rank in the list, if it differs from an env var?

JP-Ellis replied:
The CLI argument takes the priority, as it is explicitly set by the end user for a specific invocation and therefore is the _most_ specific. I'll amend the RFC to make this clear.

YOU54F replied:
okie dokie, just need to check what happens with the pact broker client. I know that for example --dry-run is implicitly set to false from the cli if not passed, should the env var overwrite that. I assumed it would, as it allows me to set an configurable env var in a CI action, that I can tweak and it can change the behaviour of the CLI, without me needing to make a code change (so I could allow a can-i-deploy step to have `PACT_BROKER_CAN_I_DEPLOY_DRY_RUN=true` and it take effect.

https://github.com/pact-foundation/pact_broker-client?tab=readme-ov-file#can-i-deploy

JP-Ellis commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,264 @@
+
+  > [!NOTE]
+  >
+  > macOS is considered a Unix-like system for the purposes of this RFC. Should an end-user prefer to use a different location (such as `~/Library/Preferences`), they may do so by setting the `XDG_CONFIG_HOME` environment variable.
```
An end-user could set either option, though they would be slightly different. The following would be equivalent:

- `XDG_CONFIG_HOME=$HOME/Library/Preferences` as the result Pact config directory becomes `~/Library/Preferences/pact` and the config file would be `~/Library/Preferences/pact/config.toml`
- `PACT_CONFIG_FILE=$HOME/Library/Preferences/pact/config.toml`

The `PACT_DATA_HOME` is not relevant when parsing config files. Though it might be worth amending the RFC such that the fallback should be `$PACT_DATA_HOME/config.toml` if `PACT_DATA_HOME` is set, and finally to fallback to `~/.pact/config.toml`.

The reason `PACT_DATA_HOME` is typically not used is that Linux and Unix-based systems separate the config from data files (e.g., you would never have a config file in `/usr/lib` or `/usr/share`, and it is always found in `/etc`). The same convention is in effect for user config files, with the files stored in `~/.config` and data files in `~/.local`.

YOU54F replied:
>The reason PACT_DATA_HOME is typically not used is that Linux and Unix-based systems separate the config from data files (e.g., you would never have a config file in /usr/lib or /usr/share, and it is always found in /etc). The same convention is in effect for user config files, with the files stored in ~/.config and data files in ~/.local.

Okay this makes sense, did a bit of background reading on macos/XDG_CONFIG_HOME with regards to ported unix tools and expected behaviour and your reasoning seems sound. I think the general expectation is someone can backup their config files without having to get a load of app data alongside it?

JP-Ellis commented:
Updated the RFC with a few changes:

- Clarifications around the following points:
  - Precedence of CLI arguments
  - Precedence of explicitly set Pact configuration files
  - When warning should _not_ be raised if the fallback location is used

When updating these, I also identified a gap in the way the TOML configuration ought to handle paths. So I have added [a subsection](https://github.com/pact-foundation/roadmap/blob/rfc/configuration-and-shared-files/rfc/0000-configuration-and-shared-files.md#path-handling) to the configuration section to make this clear.

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+- **Configuration**: Define a standard way to store runtime configuration for Pact implementations.
+- **Shared Files**: Define a standard way to store shared files across the Pact ecosystem.
+
+## Environment Variables
```
Language level:

Pact-PHP allow using `PACT_LOGLEVEL` to configure log level for FFI library.

I'm not sure if other implementations has this ability, and if they has, should we use this variable name across implementations?

JP-Ellis replied:
Logging is, in my experience, one of these things which can vary significantly based on the language's convention. So I don't know if that is something we want to fix in an RFC.

If we did implement this, my suggestion would be:

- Have `PACT_LOG` as the environment variable name, or have it stored at:

  ```
  [pact]
  log = "..."
  ```

  in the config TOML.

- Have the values be either of the following (case insensitive):
  - `DEBUG`
  - `INFO`
  - `WARN` or `WARNING`
  - `ERROR`

But like I said, I am hesitant. Logging can be complex, and I do think it is best to fit in with the conventions of the language.

tienvx replied:
I'm fine with leaving this for future RFC

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+  [pact]
+  plugin_dir = "~/.local/share/pact/plugins"
+  data_home = "~/.local/share/pact"
+  do_not_track = true
```
I feel like:

* `log_level = debug` or
* `loglevel = debug`

is important and can be included in this file.

But like `PACT_LOGLEVEL` env, I'm not sure how this config/env behave in different pact's implementations.

tienvx commented on rfc/0000-configuration-and-shared-files.md:
````diff
@@ -0,0 +1,296 @@
+  do_not_track = true
+  ```
+
+  The following entries are proposed as part of this RFC:
````
Is it only me, or Github format large of the code starting from this line with red background?

JP-Ellis replied:
If you are just reading the diff, it may not have the correct syntax highlighting. I would recommend you read the rendered view (though it does make commenting more difficult).

tienvx replied:
Nothing we can do about this. We can resolve this one.

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+
+  - `plugin_dir`: The directory where shared plugins are stored. Equivalent to `PACT_PLUGIN_DIR`.
+  - `data_home`: The directory where shared files are stored. Equivalent to `PACT_DATA_HOME`.
+  - `do_not_track`: If set to `true`, the Pact implementation _must not_ report any usage analytics data back to the Pact Foundation.
```
Should we mention the default value is `false` (opt-out) ?

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+  - `base_url`: The base URL of the Pact Broker. Equivalent to `PACT_BROKER_BASE_URL`.
+  - `username`: The username to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_USERNAME`.
+  - `password`: The password to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_PASSWORD`.
+  - `token`: The token to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_TOKEN`.
```
I'm not sure, that if username/password is set then token is ignored and vice versa?

Should we mention that if it's true?

JP-Ellis replied:
When talking about the environment variables, I do state that having a mix of username/password with token is undefined behaviour.

Having said that, this does bring a good question as to how best to handle a local override to a user config; as this could quite easily result in all three being defined.

To resolve this, I actually think it might be best to change the TOML config to:

```toml
[broker]
auth = { token = "..." }
# or
auth = { username = "...", password = "..." }
```

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+
+If the operating system cannot be detected, the fallback location _should_ be used.
+
+If the operating system is detected and the fallback location is found, a warning _should_ be displayed to inform the user to relocate the directory. This exception to this is if the location of the shared files is explicitly set, in which case the warning _must not_ be displayed.
```
These 2 sentences appear twice and I don't really understand them.

Can you explain them a bit more? Or an example for the warning message would be nice.

tienvx replied:
Nevermind, after reading again, it's basically telling me to move my `~/.pact/plugins` into `~/.local/share/pact/plugins`

tienvx commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+At present, there are only a few files which are shared across the Pact ecosystem. These are:
+
+- Plugins, currently stored in `~/.pact/plugins`
+- The Pact core library (i.e., the Pact FFI), though only in some implementations.
```
what are proposed paths? e.g. `~/.local/share/pact/ffi` and fallback `~/.pact/ffi` ?

JP-Ellis replied:
I don't actually specify how the FFI binaries/headers should be stored, mostly because I don't know what practices are used across the ecosystem.

Having `$PACT_DATA_HOME/ffi/pact.h` and `$PACT_DATA_HOME/ffi/(lib)pact.(ext)` seems quite reasonable to me though.

This assumes it is dynamically linked though, which isn't the case for all implementations.

mefellows replied:
We might need to think about versioning here. For instance, a user might be using Pact JS and Pact Go. It might be OK to have a latter version of the FFI for one and not the other (e.g. they have just started using Pact Go, and are on the latest version, but Pact JS has an existing suite and they are on an older version). In theory, this should only be a problem for major versions of the FFI, but in practice this isn't always the case.

mefellows replied:
Or perhaps a way to address it is by convention. 

e.g. 

* [base path]/ffi/pact.h
* [base path]/ffi/pact.so
* [base path]/ffi/pact-js/pact.h
* [base path]/ffi/pact-js/pact.so
* [base path]/ffi/pact-go/pact.h
* [base path]/ffi/pact-go/pact.so

Pact JS and Pact Go would have separate versions, and all other libraries use the fallback

JP-Ellis replied:
I was actually thinking more about this, and I was wondering whether we should recommend the standard Unix way of organising these files:

```
$PACT_DATA_HOME/include/pact.h
$PACT_DATA_HOME/lib/libpact.so
```

and then libraries can opt in to library-specific version of the FFI with

```
$PACT_DATA_HOME/include/pact.js.h
$PACT_DATA_HOME/lib/libpact.js.so
$PACT_DATA_HOME/include/pact.go.h
$PACT_DATA_HOME/lib/libpact.go.so
...
```

This follows the same pattern of having multiple version of libraries presented as:

```
lib/libfoo.so      -> libfoo.2.so
lib/libfoo.2.so    -> libfoo.2.1.so
lib/libfoo.2.1.so
lib/libfoo.1.so    -> libfoo.1.8.so
lib/libfoo.1.8.so
```

(the arrows are symlinks, but we need not use symlinks here)

JP-Ellis commented on rfc/0000-configuration-and-shared-files.md:
```diff
@@ -0,0 +1,296 @@
+
+  - `plugin_dir`: The directory where shared plugins are stored. Equivalent to `PACT_PLUGIN_DIR`.
+  - `data_home`: The directory where shared files are stored. Equivalent to `PACT_DATA_HOME`.
+  - `do_not_track`: If set to `true`, the Pact implementation _must not_ report any usage analytics data back to the Pact Foundation.
```
Will update!

JP-Ellis commented:
## Plan for accepting RFC

This RFC has been open for a while, and has received little additional feedback for a while, with the general consensus so far being in favour of this change.

I have a partial implementation working privately, which I will spin out into a dedicated Rust crate once this RFC is merged.

I am making some suggestions below, and therefore would like to leave this open for at least another 2-3 weeks to give everyone the opportunity to give more feedback.

## Amendment for FFI header/lib path

I would like to take this opportunity to add a standard location for the Pact library (should we use dynamic linking). Some discussion was started in [this conversation](https://github.com/pact-foundation/roadmap/pull/98#discussion_r1745865509).

I think this RFC should be amended with the following change:

---

A new optional section will be added to the config containing two values:

```
[pact.ffi]
include_path = "..."
lib_path = "..."
```

These values can also be set through the environment variables

- `PACT_FFI_INCLUDE_PATH`
- `PACT_FFI_LIBRARY_PATH

Should they be unset, they should default to:

- Searching `$PACT_DATA_HOME/include` and `$PACT_DATA_HOME/lib`
- Searching the system paths (`C_INCLUDE_PATH` and `LIBRARY_PATH` and/or `LD_LIBRARY_PATH` respectively).

Support for multiple paths is allowed by using a colon-separated environment variable, or using a list in TOML. If multiple paths are specified, then each path is tried in order.

## Suffixes (home, dir, path, etc.)

One last thing while re-reading the RFC that I noticed is that there is little consistency between `_path` or `_home` or `_dir`; or using `_file` or not.

The _only_ precedent we have at the moment in the Pact ecosystem is the `PACT_PLUGIN_DIR` environment variable.

Here are my thoughts:

- Anything that should point to a specific file should end in `_FILE` to avoid confusion with a directory or an option.
- Anything that should point to a specific directory should end in either `_DIR` or `_DIRECTORY` to avoid confusion with a file or an option
- Anything that points to a collection of directories (including a collection of one directory) should end in `_PATH`

This follows my experience with environment variables from using other tools on Linux/Unix. This change would change the `PACT_DATA_HOME` environment variable to `PACT_DATA_DIR` for example.

Please let me know if you think this is something worth adding to the RFC, or whether this is an unnecessary complication.
